File: android/app/src/main/java/com/spandan/app/camera/RoiPixelAverager.kt
package com.spandan.app.camera

import android.graphics.Rect
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageProxy
import com.spandan.app.signal.RgbSample

/**
 * Spatially averages R/G/B pixel intensities inside an ROI rect, sampled
 * directly from the raw YUV_420_888 planes of a CameraX ImageProxy -- no
 * Bitmap/JPEG round-trip needed.
 *
 * This IS real, permanent code: spatial averaging is plain arithmetic, and
 * the YUV->RGB conversion below is the standard, universal BT.601 colorspace
 * formula (a physical-sensor-format conversion, not a tuned DSP parameter),
 * so fixed constants here are fine -- unlike CHROM/POS/bandpass coefficients
 * elsewhere in this project, which must stay out of this codebase until the
 * MATLAB side finalizes them.
 */
object RoiPixelAverager {

    // Skips every other pixel in each direction -- a performance knob, not a
    // signal-processing parameter. Safe to tune without touching any "real
    // algorithm" concern.
    private const val SAMPLE_STRIDE = 2

    @ExperimentalGetImage
    fun averageRgb(imageProxy: ImageProxy, roiSensorRect: Rect): RgbSample? {
        val image = imageProxy.image ?: return null
        val width = imageProxy.width
        val height = imageProxy.height

        val left = roiSensorRect.left.coerceIn(0, width - 1)
        val top = roiSensorRect.top.coerceIn(0, height - 1)
        val right = roiSensorRect.right.coerceIn(left + 1, width)
        val bottom = roiSensorRect.bottom.coerceIn(top + 1, height)
        if (right <= left || bottom <= top) return null

        val yPlane = image.planes[0]
        val uPlane = image.planes[1]
        val vPlane = image.planes[2]
        val yBuffer = yPlane.buffer
        val uBuffer = uPlane.buffer
        val vBuffer = vPlane.buffer

        var sumR = 0L
        var sumG = 0L
        var sumB = 0L
        var count = 0

        var y = top
        while (y < bottom) {
            var x = left
            while (x < right) {
                val yIndex = y * yPlane.rowStride + x * yPlane.pixelStride
                val uvRow = y / 2
                val uvCol = x / 2
                val uIndex = uvRow * uPlane.rowStride + uvCol * uPlane.pixelStride
                val vIndex = uvRow * vPlane.rowStride + uvCol * vPlane.pixelStride

                if (yIndex < yBuffer.capacity() && uIndex < uBuffer.capacity() && vIndex < vBuffer.capacity()) {
                    val yVal = yBuffer.get(yIndex).toInt() and 0xFF
                    val uVal = (uBuffer.get(uIndex).toInt() and 0xFF) - 128
                    val vVal = (vBuffer.get(vIndex).toInt() and 0xFF) - 128

                    // Standard BT.601 YUV -> RGB.
                    val r = yVal + 1.402 * vVal
                    val g = yVal - 0.344136 * uVal - 0.714136 * vVal
                    val b = yVal + 1.772 * uVal

                    sumR += r.coerceIn(0.0, 255.0).toLong()
                    sumG += g.coerceIn(0.0, 255.0).toLong()
                    sumB += b.coerceIn(0.0, 255.0).toLong()
                    count++
                }
                x += SAMPLE_STRIDE
            }
            y += SAMPLE_STRIDE
        }

        if (count == 0) return null
        return RgbSample(
            timestampMs = System.currentTimeMillis(),
            red = sumR.toFloat() / count,
            green = sumG.toFloat() / count,
            blue = sumB.toFloat() / count
        )
    }
}

File: ios/Spandan/Sources/Camera/RoiPixelAverager.swift
import CoreVideo
import Foundation

/// Spatially averages R/G/B pixel intensities inside an ROI rect, sampled
/// directly from a bi-planar (NV12-style) `CVPixelBuffer` -- no `UIImage`/`CGImage`
/// round-trip needed. Ported from android/app/.../camera/RoiPixelAverager.kt.
///
/// Assumes `kCVPixelFormatType_420YpCbCr8BiPlanarFullRange` (the format
/// CameraController explicitly requests): plane 0 is Y (1 byte/pixel), plane 1 is
/// interleaved Cb,Cr at half resolution in each dimension (2 bytes per 2x2 luma
/// block). This is iOS's bi-planar layout, different from Android's generic
/// 3-plane `YUV_420_888` (the Kotlin source indexes 3 separate Y/U/V planes) --
/// the BT.601 full-range YUV->RGB formula itself is identical and copied exactly.
enum RoiPixelAverager {

    // Skips every other pixel in each direction -- a performance knob, not a
    // signal-processing parameter, same as the Android port's SAMPLE_STRIDE.
    private static let sampleStride = 2

    static func averageRgb(pixelBuffer: CVPixelBuffer, roi: CGRect) -> RgbSample? {
        guard CVPixelBufferGetPlaneCount(pixelBuffer) >= 2 else { return nil }

        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        guard
            let yBase = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 0),
            let uvBase = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 1)
        else { return nil }

        let yStride = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0)
        let uvStride = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 1)
        let yPtr = yBase.assumingMemoryBound(to: UInt8.self)
        let uvPtr = uvBase.assumingMemoryBound(to: UInt8.self)

        let left = max(0, min(width - 1, Int(roi.minX.rounded())))
        let top = max(0, min(height - 1, Int(roi.minY.rounded())))
        let right = max(left + 1, min(width, Int(roi.maxX.rounded())))
        let bottom = max(top + 1, min(height, Int(roi.maxY.rounded())))
        guard right > left, bottom > top else { return nil }

        var sumR = 0
        var sumG = 0
        var sumB = 0
        var count = 0

        var y = top
        while y < bottom {
            var x = left
            while x < right {
                let yIndex = y * yStride + x
                let uvRow = y / 2
                let uvCol = (x / 2) * 2 // interleaved Cb,Cr pair start
                let uvIndex = uvRow * uvStride + uvCol

                let yVal = Int(yPtr[yIndex])
                let uVal = Int(uvPtr[uvIndex]) - 128
                let vVal = Int(uvPtr[uvIndex + 1]) - 128

                // Standard BT.601 YUV -> RGB (same formula as the Android port).
                let r = Double(yVal) + 1.402 * Double(vVal)
                let g = Double(yVal) - 0.344136 * Double(uVal) - 0.714136 * Double(vVal)
                let b = Double(yVal) + 1.772 * Double(uVal)

                sumR += Int(min(255.0, max(0.0, r)))
                sumG += Int(min(255.0, max(0.0, g)))
                sumB += Int(min(255.0, max(0.0, b)))
                count += 1

                x += sampleStride
            }
            y += sampleStride
        }

        guard count > 0 else { return nil }
        return RgbSample(
            timestampMs: Int64(Date().timeIntervalSince1970 * 1000),
            red: Float(sumR) / Float(count),
            green: Float(sumG) / Float(count),
            blue: Float(sumB) / Float(count)
        )
    }
}
